### PR TITLE
scale to zero with cpu/mem scaler

### LINKED ISSUE #4269: Support scaling to Zero with combine of Cron and cpu/mem scaler
### Proposal

My scaled object
```
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: app
spec:
  maxReplicaCount: 10
  minReplicaCount: 0
  scaleTargetRef:
    kind: Deployment
    name: app
  triggers:
    - metadata:
        desiredReplicas: "2"
        end: 0 21 * * *
        start: 30 7 * * *
        timezone: Asia/Ho_Chi_Minh
      type: cron
    - metadata:
        value: "85"
      metricType: Utilization
      type: cpu
```

I would like to scale my deployment to Zero but the cpu/mem scaler prevents that

Proposal from @zroubalik: `we can allow scale to 0, in case there are multiple triggers not just cpu/mem and the activation will be driven by the non-cpu/mem scaler.`

### Use-Case

_No response_

### Is this a feature you are interested in implementing yourself?

No

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by gauron99**
add option to scale to zero with cpu/mem triggers when:
- SO has atleast one external trigger

if SO has no external triggers, only cpu/mem triggers AND minReplicas is not set or is 0 - return error in so creation/validation

- added exception for gci because of https://github.com/kedacore/keda/issues/4379 - waiting for resolution

### Checklist

(https://github.com/kedacore/governance/blob/main/SCALERS.md)
- [x] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [x] Tests have been added
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))
- [x] Update docs

docs PR: https://github.com/kedacore/keda-docs/pull/1088

Fixes #4269

### CODE REVIEW

**Comment by gauron99** [comment]
try running the test again pls :pray: Ive finalized it

**Comment by gauron99** [comment]
sorry i had to redo the tests a few times, example-hpa image doesnt work for me in ocp for some reason when im trying to test it so im doing it kind of blind. It should be fixed now, if you could please run both tests again :pray:

**Review by zroubalik — COMMENTED** [review]
In the test output I can read: 
```
No test has been executed, please review your regex: 'memory*
```

it is strange though 🤷‍♂️

**Comment by gauron99** [comment]
It looks like there was some whitespace characters (newlines?) after the `/run-e2e memory*` text which messed up with the regex matching - i made a PR https://github.com/kedacore/keda/pull/4363

**Comment by JorTurFer** [comment]
> It looks like there was some whitespace characters (newlines?) after the `/run-e2e memory*` text which messed up with the regex matching - i made a PR #4363

Yeah, it's a known bug that we forget from time to time xD 
If the comment has a new line, the regex fails

**Comment by tomkerkhove** [comment]
> > It looks like there was some whitespace characters (newlines?) after the `/run-e2e memory*` text which messed up with the regex matching - i made a PR #4363
> 
> Yeah, it's a known bug that we forget from time to time xD If the comment has a new line, the regex fails

Jorge has to always remind me about this because I'm sloppy - Sorry!

File: cmd/webhooks/main.go [hunk 1]
```diff
@@ -33,7 +33,6 @@ import (
 	kedav1alpha1 "github.com/kedacore/keda/v2/apis/keda/v1alpha1"
 	"github.com/kedacore/keda/v2/pkg/k8s"
 	kedautil "github.com/kedacore/keda/v2/pkg/util"
-	//+kubebuilder:scaffold:imports
```
Comment (zroubalik) [inline cmd/webhooks/main.go:36]: why is this change here?

File: cmd/operator/main.go [hunk 2]
```diff
@@ -42,7 +42,6 @@ import (
 	"github.com/kedacore/keda/v2/pkg/metricsservice"
 	"github.com/kedacore/keda/v2/pkg/scaling"
 	kedautil "github.com/kedacore/keda/v2/pkg/util"
-	//+kubebuilder:scaffold:imports
```
[the remark this answers is not in the document]
Comment (gauron99) [reply cmd/operator/main.go:45]: all of these were suggestions from lint, without them the static check didnt pass

[the remark this answers is not in the document]
Comment (zroubalik) [reply cmd/operator/main.go:45]: even if you put empty line before this one?

File: CHANGELOG.md [hunk 3]
```diff
@@ -42,7 +42,7 @@ To learn more about active deprecations, we recommend checking [GitHub Discussio
 - [v1.0.0](#v100)
 
 ## Unreleased
-
+- **cpu/mem scaler**: Add option to scale to zero with cpu/mem scaler([#4269](https://github.com/kedacore/keda/issues/4269))
```
Comment (zroubalik) [inline CHANGELOG.md:45]:
```suggestion
- **CPU/Memory scaler**: Add support for scale to zero if there are multiple triggers ([#4269](https://github.com/kedacore/keda/issues/4269))
```
Move this to the New section

File: apis/keda/v1alpha1/scaledobject_webhook.go [hunk 4]
```diff
@@ -138,6 +141,7 @@ func verifyHpas(incomingSo *ScaledObject, action string) error {
 	return nil
 }
 
+// TODO
```
Comment (zroubalik) [inline apis/keda/v1alpha1/scaledobject_webhook.go:144]: ?

File: apis/keda/v1alpha1/scaledobject_webhook.go [hunk 5]
```diff
@@ -238,6 +242,24 @@ func verifyCPUMemoryScalers(incomingSo *ScaledObject, action string) error {
...
+			scaleToZeroErr := true
+			for _, trig := range incomingSo.Spec.Triggers {
+				if trig.Type != cpuString && trig.Type != memoryString {
+					scaleToZeroErr = false
+					break
+				}
+			}
+
+			if (scaleToZeroErr && incomingSo.Spec.MinReplicaCount == nil) || (scaleToZeroErr && *incomingSo.Spec.MinReplicaCount == 0) {
+				err := fmt.Errorf("scaledobject has no external triggers AND minReplica is 0 (it can never scale down to 0)")
```
Comment (zroubalik) [inline apis/keda/v1alpha1/scaledobject_webhook.go:258]:
```suggestion
				err := fmt.Errorf("scaledobject has only cpu/memory triggers AND minReplica is 0 (scale to zero doesn't work in this case)")
```

File: pkg/scaling/scale_handler.go [hunk 6]
```diff
@@ -566,9 +566,19 @@ func (h *scaleHandler) getScaledObjectState(ctx context.Context, scaledObject *k
 		}
 
 		for _, spec := range metricSpecs {
-			// skip cpu/memory resource scaler, these scalers are also always Active
+			// if cpu/memory resource scaler has minReplicas==0 & atleast one external
```
Comment (zroubalik) [inline pkg/scaling/scale_handler.go:569]:
```suggestion
			// if cpu/memory resource scaler has minReplicas==0 & at least one external
```

File: apis/keda/v1alpha1/scaledobject_webhook.go (hunk 4 above)
Comment (gauron99) [reply apis/keda/v1alpha1/scaledobject_webhook.go:144]: leftover comment, mb

**Review by zroubalik — APPROVED** [review]
LGTM, thanks!

**Comment by JorTurFer** [comment]
I'll check this during the day

File: .golangci.yml [hunk 7]
```diff
@@ -107,6 +107,21 @@ issues:
...
+    - path: cmd/webhooks/main.go
+      linters:
+        - gci
+    - path: controllers/keda/suite_test.go
+      linters:
+        - gci
+    - path: apis/keda/v1alpha1/scaledobject_webhook_test.go
+      linters:
+        - gci
+
```
Comment (JorTurFer) [inline .golangci.yml:124]: Why is this required now?

File: pkg/scaling/scale_handler.go [hunk 8]
```diff
@@ -566,9 +566,19 @@ func (h *scaleHandler) getScaledObjectState(ctx context.Context, scaledObject *k
...
+			// trigger exists -> object can be scaled to zero
+			//
 			if spec.External == nil {
-				isScaledObjectActive = true
+				cpuMemCount := 0
+				for _, trigger := range scaledObject.Spec.Triggers {
+					if trigger.Type == "cpu" || trigger.Type == "memory" {
+						cpuMemCount++
+					}
+				}
```
Comment (JorTurFer) [inline pkg/scaling/scale_handler.go:578]: Shouldn't we calculate this before the loop? I mean, if we have 3 external metrics we will evaluate the `cpuMemCount` value 3 times

**Review by JorTurFer — COMMENTED** [review]
Looking good. Some comments inline

File: .golangci.yml (hunk 7 above)
Comment (gauron99) [reply .golangci.yml:124]: the gci check wants to remove the comments - i tried to move them around but doesnt work, im tracking the issue within gci here https://github.com/kedacore/keda/issues/4379

File: pkg/scaling/scale_handler.go (hunk 8 above)
Comment (gauron99) [reply pkg/scaling/scale_handler.go:578]: its in the `if spec.External == nil` branch so whenever its non external trigger the `for` gets triggered. but if we have multiple cpu/mem triggers it would evaluate multiple times, thats true. I can move it to the beginning of the function which would mean it would evaluate even if there were no cpu/mem triggers. 
What you want me to do?

Comment (JorTurFer) [reply pkg/scaling/scale_handler.go:578]: I guess that it's better evaluating it once always than X amount of times in edge cases. 
Moving it before it's O(n) and inside the loop is O(n2)

**Review by JorTurFer — APPROVED** [review]
LGTM!
